### PR TITLE
Simplify redundant SPDX choices

### PR DESCRIPTION
**Pull request by sschuberth**
Please have a look at the individual commit messages for the details.

### CODE REVIEW

File: utils/spdx/src/test/kotlin/SpdxCompoundExpressionTest.kt [hunk 1]
```diff
@@ -68,7 +68,7 @@ class SpdxCompoundExpressionTest : WordSpec({
             ).toString()
         }
 
-        "create a single expression for equal operands" {
+        "create a single expression for equal AND-operands" {
```
Comment (mnonnenmacher) [inline utils/spdx/src/test/kotlin/SpdxCompoundExpressionTest.kt:71]: Commit message nit: Typo in "simplifying"

File: utils/spdx/src/main/kotlin/SpdxExpression.kt [hunk 2]
```diff
@@ -300,6 +300,8 @@ class SpdxCompoundExpression(
     override fun validChoices(): Set<SpdxExpression> =
         when (operator) {
             SpdxOperator.AND -> {
+                // If there is a top-level `AND` in an expressions, create the combinations of all choices on the left
```
Comment (mnonnenmacher) [inline utils/spdx/src/main/kotlin/SpdxExpression.kt:305]: "expressions" -> "expression"

File: utils/spdx/src/main/kotlin/SpdxExpression.kt [hunk 3]
```diff
@@ -317,7 +319,11 @@ class SpdxCompoundExpression(
                 }
             }
 
-            SpdxOperator.OR -> children.flatMapTo(mutableSetOf()) { it.validChoices() }
+            SpdxOperator.OR -> {
+                // If there is a top-level `OR` in an expressions, the operands already are the choices and just need to
```
Comment (mnonnenmacher) [inline utils/spdx/src/main/kotlin/SpdxExpression.kt:325]: "expressions" -> "expression"

File: utils/spdx/src/test/kotlin/SpdxCompoundExpressionTest.kt [hunk 4]
```diff
@@ -93,5 +93,12 @@ class SpdxCompoundExpressionTest : WordSpec({
             // Compare string representations to not rely on semantic equality.
             expression.simplify().toString() shouldBe "MIT"
         }
+
+        "remove redundant choices" {
+            val expression = "Apache-2.0 AND (Apache-2.0 OR MIT) AND MIT".toSpdx()
+
+            // Compare string representations to not rely on semantic equality.
```
Comment (mnonnenmacher) [inline utils/spdx/src/test/kotlin/SpdxCompoundExpressionTest.kt:100]: Should this say "structural equality" instead? I'm actually not sure if I understand the comment.

File: utils/spdx/src/main/kotlin/SpdxExpression.kt [hunk 5]
```diff
@@ -261,7 +261,9 @@ class SpdxCompoundExpression(
             }
         }
 
-        return flattenedChildren.singleOrNull() ?: SpdxCompoundExpression(operator, flattenedChildren)
+        val flattened = flattenedChildren.singleOrNull() ?: SpdxCompoundExpression(operator, flattenedChildren)
+        val choices = flattened.validChoices()
```
Comment (mnonnenmacher) [inline utils/spdx/src/main/kotlin/SpdxExpression.kt:265]: As you added explanatory comments in the previous commit I think one here would also be nice to explain why the choices are taken into account.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: utils/spdx/src/test/kotlin/SpdxCompoundExpressionTest.kt (hunk 4 above)
Comment (sschuberth) [reply utils/spdx/src/test/kotlin/SpdxCompoundExpressionTest.kt:108]: As this is about simplifying expressions, I want to be sure that e.g. `MIT AND MIT` is simplified to exactly `MIT` although it is `"MIT AND MIT".toSpdx() == "MIT".toSpdx()`.

**Comment by sschuberth** [comment]
PS: It occurred to me that we probably could convert several functions, like `decompose()`, to lazy properties to avoid recomputation.

File: utils/spdx/src/test/kotlin/SpdxCompoundExpressionTest.kt (hunk 4 above)
Comment (mnonnenmacher) [reply utils/spdx/src/test/kotlin/SpdxCompoundExpressionTest.kt:108]: Yes, but isn't "MIT AND MIT" semantically equal with "MIT", but not syntactically?

**Review by mnonnenmacher — APPROVED** [review]

File: utils/spdx/src/test/kotlin/SpdxCompoundExpressionTest.kt (hunk 4 above)
Comment (sschuberth) [reply utils/spdx/src/test/kotlin/SpdxCompoundExpressionTest.kt:108]: Yes, exactly. The comment is meaning to say: "If we would not use string-comparison here, expressions would be compared for being semantically equal, which is not what we want."

**Pull request merged**